src/gui/sdlmain.cpp

```diff
@@ -3386,7 +3387,132 @@ static void set_output(Section *sec, bool should_stretch_pixels)
 	SDL_SetWindowOpacity(sdl.window, alpha);
 }
 
-//extern void UI_Run(bool);
+static std::optional<SDL_Surface *> get_rendered_surface()
+{
+	// Variables common to all screen-modes
+	const auto renderer = SDL_GetRenderer(sdl.window);
+	const auto canvas   = get_canvas_size(sdl.desktop.type);
+
+#if C_OPENGL
+	// Get the OpenGL-renderer surface
+	// -------------------------------
+	if (sdl.desktop.type == SCREEN_OPENGL) {
+		// Setup our OpenGL image properties
+		constexpr int gl_channels       = 3; // RBG (no alpha)
+		constexpr int gl_bits_per_pixel = gl_channels * 8; // 8-bpp
+		const size_t bytes_per_row      = gl_channels * canvas.w;
+
+		// Allocate a 24-bit surface to be populated
+		const auto surface = SDL_CreateRGBSurface(SDL_SWSURFACE,
+		                                          canvas.w,
+		                                          canvas.h,
+		                                          gl_bits_per_pixel,
+		                                          RMASK,
+		                                          GMASK,
+		                                          BMASK,
+		                                          AMASK);
+		if (!surface) {
+			LOG_WARNING("SDL: Failed creating a surface for OpenGL because %s",
+			            SDL_GetError());
+			return {};
+		}
+		// Per OpenGL's documentation:
+		// The glReadPixels function starts at the lower left corner: (x
+		// + i, y + j) and iterates for 0 <= i < width and 0 <= j <
+		// height. It describes the pxiels as being "the i'th pixel in
+		// the j'th row". Pixels are returned in row-order from the
+		// lowest to the highest row, left to right in each row.
+		std::vector<uint8_t> pixels(bytes_per_row * canvas.h);
+		glReadPixels(0, 0, canvas.w, canvas.h, GL_RGB, GL_UNSIGNED_BYTE, pixels.data());
```

Comment:
This approach certainly works, but it's quite slow (just Google "glReadPixels very slow"). Taking screenshots using Cmd+F5 is an instantaneous operation on my Mac, but when I use Opt+F5, DOSBox freezes for about 1-1.5 second. To make sure it's not the PNG compression slowing things down, I set it to 0 (no compression), but that hasn't changed anything.

A more efficient way of doing this would be to render the final output image into a framebuffer first, then that can be fed to the PNG compressor much more efficiently. But this would require changing the graphics pipeline a bit, so not asking you to do that because that would bleed into the graphics rendering rework task. Just a note for the future, I guess! 😄

Reply:
To elaborate further, rendering the emulated computer's output and the OSD into separate framebuffers (textures) would allow us to do stuff like this. This is where we want to get to, eventually.

![image](https://user-images.githubusercontent.com/698770/200964693-eaaa9204-7357-45ae-ba32-6f7d888f8fc6.png)

Reply:
(resolving to go ahead w/ the merge - not because this comment is resolved.  I've added a comment to the PR itself below referencing this ticket and asking for help)